# 1 Assignment 3

In [61]:
import os
import os
import pickle
import pandas as pd
import numpy as np
import urllib.request

1. Import the `REC0111.sav`, `RE223132.sav` and `RE516171.sav` files and their variables and values labels from this path `"../../_data/endes/2019"`. The name of imported files should be named as `rec_1`, `rec_2` and `rec_3` for files `REC0111.sav`, `RE223132.sav` and `RE516171.sav` respectively. The name of the variable and value labels should be `var_labels1` and `value_labels1` for `rec1`, `var_labels2` and `value_labels2` for `rec2`, and `var_labels3` and `value_labels3` for `rec3`. **Hint: See the section 3.3.4 of [the lecture 3](https://github.com/alexanderquispe/Diplomado_PUCP/blob/main/Lecture_3/Lecture_3.ipynb)**

In [62]:
#pwd

In [63]:
#

In [64]:
#pwd

In [65]:
%pip install pyreadstat
import pyreadstat
import pandas as pd

rec_1 = pd.read_spss(r"../../_data/endes/2019/REC0111.sav")
rec_2 = pd.read_spss(r"../../_data/endes/2019/RE223132.sav")
rec_3 = pd.read_spss(r"../../_data/endes/2019/RE516171.sav")

rec_1

Note: you may need to restart the kernel to use updated packages.


ID1             HHID              CASEID    V001  V002  V003  \
0      2019.0        000100201        000100201  2     1.0   2.0   2.0   
1      2019.0        000100201        000100201  3     1.0   2.0   3.0   
2      2019.0        000102801        000102801  2     1.0  28.0   2.0   
3      2019.0        000102801        000102801  6     1.0  28.0   6.0   
4      2019.0        000104801        000104801  2     1.0  48.0   2.0   
...       ...              ...                 ...     ...   ...   ...   
38330  2019.0        325406201        325406201  2  3254.0  62.0   2.0   
38331  2019.0        325406301        325406301  2  3254.0  63.0   2.0   
38332  2019.0        325407001        325407001  2  3254.0  70.0   2.0   
38333  2019.0        325407201        325407201  2  3254.0  72.0   2.0   
38334  2019.0        325407401        325407401  2  3254.0  74.0   2.0   

         V004    V007    V008  V009  ...  QD333_4  QD333_5  QD333_6  UBIGEO  \
0         1.0  2019.0  1434.0   4.0  ...       No       No       No  010101   
1         1.0  2019.0  1434.0   1.0  ...       No       No       No  010101   
2         1.0  2019.0  1434.0   6.0  ...       No       No       No  010101   
3         1.0  2019.0  1434.0   3.0  ...       No       No       No  010101   
4         1.0  2019.0  1434.0   5.0  ...       No       No       No  010101   
...       ...     ...     ...   ...  ...      ...      ...      ...     ...   
38330  3254.0  2019.0  1440.0  12.0  ...       No       No       No  250401   
38331  3254.0  2019.0  1440.0   6.0  ...       No       No       No  250401   
38332  3254.0  2019.0  1440.0   7.0  ...       No       No       No  250401   
38333  3254.0  2019.0  1440.0  12.0  ...       No       No       No  250401   
38334  3254.0  2019.0  1440.0  10.0  ...       No       No       No  250401   

        V022      V005          V190      V191  \
0        3.0  154803.0          Rico  1.234450   
1        3.0  154803.0          Rico  1.234450   
2        3.0  154803.0          Rico  1.295611   
3        3.0  154803.0          Rico  1.295611   
4        3.0  154803.0        Pobrer -0.256431   
...      ...       ...           ...       ...   
38330  249.0  244995.0  El más pobre -1.750187   
38331  249.0  244995.0  El más pobre -1.676861   
38332  249.0  459792.0  El más pobre -1.585333   
38333  249.0  459792.0  El más pobre -1.650159   
38334  249.0  244995.0  El más pobre -1.644720   

                                        mujeres12a49  NCONGLOME  
0                    Mujeres de 15 a 49 años de edad     7065.0  
1      Mujeres de 12 a 14 de edad, nunca embarazadas     7065.0  
2                    Mujeres de 15 a 49 años de edad     7065.0  
3                    Mujeres de 15 a 49 años de edad     7065.0  
4                    Mujeres de 15 a 49 años de edad     7065.0  
...                                              ...        ...  
38330                Mujeres de 15 a 49 años de edad    15783.0  
38331                Mujeres de 15 a 49 años de edad    15783.0  
38332                Mujeres de 15 a 49 años de edad    15783.0  
38333                Mujeres de 15 a 49 años de edad    15783.0  
38334                Mujeres de 15 a 49 años de edad    15783.0  

[38335 rows x 105 columns]

In [66]:
import pyreadstat

file_path = r"../../_data/endes/2019/REC0111.sav"
df, meta = pyreadstat.read_sav(file_path, apply_value_formats=True)

# Extract variable labels
if isinstance(meta.column_labels, list):
    var_labels1 = {col: label for col, label in zip(df.columns, meta.column_labels)}
else:
    var_labels1 = meta.column_labels

# Extract value labels
value_labels1 = meta.variable_value_labels

In [67]:
import pyreadstat

file_path = r"../../_data/endes/2019/RE223132.sav"
df, meta = pyreadstat.read_sav(file_path, apply_value_formats=True)

# Extract variable labels
if isinstance(meta.column_labels, list):
    var_labels2 = {col: label for col, label in zip(df.columns, meta.column_labels)}
else:
    var_labels2 = meta.column_labels

# Extract value labels
value_labels2 = meta.variable_value_labels

In [68]:
import pyreadstat

file_path = r"../../_data/endes/2019/RE516171.sav"
df, meta = pyreadstat.read_sav(file_path, apply_value_formats=True)

# Extract variable labels
if isinstance(meta.column_labels, list):
    var_labels3 = {col: label for col, label in zip(df.columns, meta.column_labels)}
else:
    var_labels3 = meta.column_labels

# Extract value labels
value_labels3 = meta.variable_value_labels


2. Select the following columns for each data set:
|Data|Columns|
|---|---|
|rec1| CASEID, V000, V001, V002, V003, V004, V007, V008, V009, V010, V011, V012, V024, V102, V120, V121, V122, V123, V124, V125, V127, V133 |
|rec2| CASEID, V201, V218, V301, V302, V323, V323A, V325A, V326, V327, V337, V359, V360, V361, V362, V363, V364, V367, V372, V372A, V375A, V376, V376A, V379, V380 |
|rec3| CASEID, V501, V502, V503, V504, V505, V506, V507, V508, V509, V510, V511, V512, V513, V525, V613, V714, V715 |


Additioanlly, you should update the variables and value labels objects. They must have information only for the selected columns. The new dataframes should be name as `rec1_1`, `rec2_1`, and `rec3_1`. The new varible labels objects should be named as `new_var_labels1`, `new_var_labels2`, and `new_var_labels3`. The new value labels objects should be named as `new_value_labels1`, `new_value_labels2`, and `new_value_labels3` **Hint: Use the `loc` and column names to filter. Update the dictionary.**

In [69]:
# Selecting the columns for rec_1, rec_2, rec_3

selected_columns_rec1 = ['CASEID', 'V000', 'V001', 'V002', 'V003', 'V004', 'V007', 'V008', 'V009', 'V010', 'V011', 'V012', 'V024', 'V102', 'V120', 'V121', 'V122', 'V123', 'V124', 'V125', 'V127', 'V133']
selected_columns_rec2 = ['CASEID', 'V201', 'V218', 'V301', 'V302', 'V323', 'V323A', 'V325A', 'V326', 'V327', 'V337', 'V359', 'V360', 'V361', 'V362', 'V363', 'V364', 'V367', 'V372', 'V372A', 'V375A', 'V376', 'V376A', 'V379', 'V380']
selected_columns_rec3 = ['CASEID', 'V501', 'V502', 'V503', 'V504', 'V505', 'V506', 'V507', 'V508', 'V509', 'V510', 'V511', 'V512', 'V513', 'V525', 'V613', 'V714', 'V715']


# Updating rec_1, rec_2, rec_3 with selected columns
rec1_1 = rec_1.loc[:, selected_columns_rec1]
rec2_1 = rec_2.loc[:, selected_columns_rec2]
rec3_1 = rec_3.loc[:, selected_columns_rec3]

# Updating variable labels for rec_1
new_var_labels1 = {key: value for key, value in var_labels1.items() if key in selected_columns_rec1}

# Updating value labels for rec_1
new_value_labels1 = {key: value for key, value in value_labels1.items() if key in selected_columns_rec1}

# Updating variable labels for rec_1
new_var_labels2 = {key: value for key, value in var_labels2.items() if key in selected_columns_rec2}

# Updating value labels for rec_1
new_value_labels2 = {key: value for key, value in value_labels2.items() if key in selected_columns_rec2}

# Updating variable labels for rec_1
new_var_labels3 = {key: value for key, value in var_labels3.items() if key in selected_columns_rec3}

# Updating value labels for rec_1
new_value_labels3 = {key: value for key, value in value_labels3.items() if key in selected_columns_rec3}

3. Generate a new column for `rec1_1` named as `year`. It should be equal to `2019`. Also, you must update this new variable for the `var_labels` dictionary. Generate a new key for `new_var_labels1` and the value for this key should be **"Year of the survey"** **Hint: Use `loc` and `update` method.**

In [70]:
# Adding a new column "year" with a constant value of 2019 to rec1_1 using loc
rec1_1.loc[:, 'year'] = 2019

# Updating variable labels for rec1_1
new_var_labels1.update({'year': 'Year of the survey'})

In [71]:
new_var_labels1

{'CASEID': 'Identificación Cuestionario Individual',
 'V001': 'Conglomerado',
 'V002': 'Número de vivienda',
 'V003': 'Número de línea de entrevistada',
 'V004': 'Unidad de área final',
 'V007': 'Año de la entrevista',
 'V008': 'Fecha de la entrevista, Codificación centenaria de meses (CMC)',
 'V009': 'Mes de nacimiento de la entrevistada',
 'V010': 'Año de nacimiento de la entrevistada',
 'V011': 'Fecha de nacimiento, Codificación centenaria de meses (CMC)',
 'V012': 'Edad actual - entrevistada',
 'V024': 'Región',
 'V000': 'Código y fase del país',
 'V102': 'Tipo de lugar de residencia',
 'V120': 'En su hogar tiene: radio',
 'V121': 'En su hogar tiene: televisión',
 'V122': 'En su hogar tiene: refrigerador',
 'V123': 'En su hogar tiene: bicicleta',
 'V124': 'En su hogar tiene: motocicleta/motocar',
 'V125': 'En su hogar tiene: coche/camión',
 'V127': 'Material predominante del piso de la vivienda',
 'V133': 'Educación en años simples',
 'year': 'Year of the survey'}

4. Merge `rec1_1`, `rec2_1`, and `rec3_1` using **CASEID**. Name this new object as `endes_2019`. **Hint: Use [this link](https://stackoverflow.com/questions/53645882/pandas-merging-101)**

In [72]:
# We are going to use the "outer" criteria to merge.
# So, we are going to have keys (CASEID) from the 3 datasets.
# NaNs are inserted for missing rows in both.

#First, we start with rec1_1 and rec2_1:
endes_2019 = rec1_1.merge(rec2_1, on='CASEID', how='outer')

In [73]:
#Second, we merge rec3_1:
endes_2019 = endes_2019.merge(rec3_1, on='CASEID', how='outer')

In [74]:
endes_2019

CASEID V000    V001  V002  V003    V004    V007    V008  \
0            000100201  2  PE6     1.0   2.0   2.0     1.0  2019.0  1434.0   
1            000100201  3  PE6     1.0   2.0   3.0     1.0  2019.0  1434.0   
2            000102801  2  PE6     1.0  28.0   2.0     1.0  2019.0  1434.0   
3            000102801  6  PE6     1.0  28.0   6.0     1.0  2019.0  1434.0   
4            000104801  2  PE6     1.0  48.0   2.0     1.0  2019.0  1434.0   
...                   ...  ...     ...   ...   ...     ...     ...     ...   
38330        325406201  2  PE6  3254.0  62.0   2.0  3254.0  2019.0  1440.0   
38331        325406301  2  PE6  3254.0  63.0   2.0  3254.0  2019.0  1440.0   
38332        325407001  2  PE6  3254.0  70.0   2.0  3254.0  2019.0  1440.0   
38333        325407201  2  PE6  3254.0  72.0   2.0  3254.0  2019.0  1440.0   
38334        325407401  2  PE6  3254.0  74.0   2.0  3254.0  2019.0  1440.0   

       V009    V010  ...    V508    V509                    V510  V511  V512  \
0       4.0  1986.0  ...  2008.0  1297.0         Tiene mes y año  21.0  11.0   
1       1.0  2007.0  ...     NaN     NaN                     NaN   NaN   NaN   
2       6.0  1983.0  ...  2011.0  1344.0         Tiene mes y año  28.0   7.0   
3       3.0  1970.0  ...  1984.0  1016.0  Tiene año - imputa mes  14.0  34.0   
4       5.0  1991.0  ...  2009.0  1320.0         Tiene mes y año  18.0   9.0   
...     ...     ...  ...     ...     ...                     ...   ...   ...   
38330  12.0  1971.0  ...  1985.0  1032.0         Tiene mes y año  14.0  34.0   
38331   6.0  1988.0  ...  2002.0  1236.0         Tiene mes y año  14.0  17.0   
38332   7.0  1973.0  ...  1986.0  1043.0         Tiene mes y año  13.0  33.0   
38333  12.0  1994.0  ...  2008.0  1301.0         Tiene mes y año  13.0  11.0   
38334  10.0  1996.0  ...  2013.0  1362.0         Tiene mes y año  16.0   6.0   

                   V513  V525 V613 V714  V715  
0       De 10 a 14 años  17.0  2.0   Sí  11.0  
1                   NaN   NaN  NaN  NaN   NaN  
2         De 5 a 9 años  18.0  3.0   Sí  14.0  
3      De 30 a más años  14.0  0.0   No   6.0  
4         De 5 a 9 años  15.0  2.0   No   6.0  
...                 ...   ...  ...  ...   ...  
38330  De 30 a más años  12.0  3.0   No   5.0  
38331   De 15 a 19 años  12.0  3.0   No   7.0  
38332  De 30 a más años  12.0  4.0   No   5.0  
38333   De 10 a 14 años  13.0  2.0   No  11.0  
38334     De 5 a 9 años  15.0  2.0   No  11.0  

[38335 rows x 64 columns]

We can verify that the number of columns is equal to the total sum of the 3 datasets columns (equal to 64).
And also check that at least this dataset has the number of rows equal to 38,335, which is the number of rows of tha largest dataset merged (rec1_1)

5. Unify all the `new_var_labels` in one object and `new_value_labels` in another one object. Name these two objects as `var_labels` and `value_labels`. Use them to generate new attributes for `endes_2019`. These attributes should be named as `var_labels` and `value_labels`. **Hint: Use `update` method.**

In [75]:
# Firstly, we are going to create an new dictionary named var_labels,
# which the information of new_var_labels1:
var_labels = new_var_labels1

# Then we use the update method to add the information from new_var_labels2:
var_labels.update(new_var_labels2)

# Finally, we are adding the information from new_var_labels3:
var_labels.update(new_var_labels3)

In [76]:
len(var_labels)
# The lenght is equal to the total number of columns of the endes_2019 dataset.

64

In [77]:
# As we did for the labels, we are going to do the same with the values:
# Create an new dictionary named value_labels:
value_labels = new_value_labels1

# Update with new_value_labels2:
value_labels.update(new_value_labels2)

# # Update with new_value_labels3:
value_labels.update(new_value_labels3)

In [78]:
# The lenght is equal to the total number of values from the 3 original datasets.
len(value_labels)

44

6. Get the min, max, sd, n_obs, n_missing for the following columns **total children ever born (V201)**, **Ideal number of children (V613)**, **Husbands education-single yrs (V715)**, and **Age at first marriage (V511)**. We want a dataframe with the following columns **Variables, Min, Max, Mean, N_obs, N_missing** and sort by the number of missing rows. **Hint: Use `describe` and `pivot` methods**

In [79]:
# We are selecting the columns of interest:
dataframe = endes_2019[["V201","V715","V613","V511"]]

# Then, we are going to create a variable with the description of the variables:
dataframe1 = dataframe.describe(include = 'all')

#Finally, we are adding the indexes "n_obs" and "n_missing":
dataframe1.loc['n_obs'] = len(dataframe)  
dataframe1.loc['n_missing'] = dataframe.isnull().sum()

In [80]:
dataframe1

V201     V715     V613          V511
count      36922.000000  25881.0  33311.0  25881.000000
unique              NaN     19.0     15.0           NaN
top                 NaN     11.0      2.0           NaN
freq                NaN   9442.0  16983.0           NaN
mean           1.781946      NaN      NaN     20.463738
std            1.749109      NaN      NaN      4.858358
min            0.000000      NaN      NaN     10.000000
25%            0.000000      NaN      NaN     17.000000
50%            1.000000      NaN      NaN     19.000000
75%            3.000000      NaN      NaN     23.000000
max           13.000000      NaN      NaN     48.000000
n_obs      38335.000000  38335.0  38335.0  38335.000000
n_missing   1413.000000  12454.0   5024.0  12454.000000

We see that the dataframe has all the indexes as rows, so we have to pivot them and only select: Min, Max, Mean, N_obs, N_missing.

In [81]:
# We are going to select the required indexes.
# and change the name of the selected index to "Variables" as requested:
dataframe2 = dataframe1.loc[["min", "max", "std", "mean", "count", "n_missing"]].reset_index().rename(columns = {'index':'Variables'})

# Now, we are going to pivot the dataframe 
# Rename the columns: adding capital letters, changing the column count to "N_obs".
# Also, we sort the dataframe according to number of missing rows (N_missing). 
dataframe2 = dataframe2.pivot_table(columns= 'Variables').rename(columns=str.capitalize).sort_values(by= 'N_missing').rename(columns = {'Count':'N_obs'})

dataframe2

Variables    N_obs   Max       Mean   Min  N_missing       Std
V201       36922.0  13.0   1.781946   0.0     1413.0  1.749109
V613       33311.0   NaN        NaN   NaN     5024.0       NaN
V511       25881.0  48.0  20.463738  10.0    12454.0  4.858358
V715       25881.0   NaN        NaN   NaN    12454.0       NaN

In [82]:
# Finally, we modifify the order of the columns as follows:
dataframe2 = dataframe2[["Min","Max","Mean","N_obs","N_missing"]]
dataframe2

Variables   Min   Max       Mean    N_obs  N_missing
V201        0.0  13.0   1.781946  36922.0     1413.0
V613        NaN   NaN        NaN  33311.0     5024.0
V511       10.0  48.0  20.463738  25881.0    12454.0
V715        NaN   NaN        NaN  25881.0    12454.0

7. Use `endes_2019` data to generate a new object named `mean_key_vars` to find the mean of **total children ever born (V201)**, **Ideal number of children (V613)**, **Husbands education-single yrs (V715)**, and **Age at first marriage (V511)** by year and department **(V024)**. Name these columns as **mean_total_children, mean_ideal_children, mean_hb_yr_educ and mean_first_marriage**, respectively. **Hint: Use groupby and [this link](https://stackoverflow.com/questions/40901770/is-there-a-simple-way-to-change-a-column-of-yes-no-to-1-0-in-a-pandas-dataframe).**

In [96]:
#First, we need to explore the nature of the chosen variables 
var = endes_2019.loc[:,[ "V201", "V613", "V715", "V511","year","V024" ]]
var.dtypes

V201     float64
V613    category
V715    category
V511     float64
year       int64
V024    category
dtype: object

In [98]:
#Second, we transform a numeric the variables that will work with
var[["V715", "V613"]] = var[['V613','V715']].apply(pd.to_numeric, errors = 'coerce' )

In [99]:
# In this problem is request to using "groupby" to sort by year and department
mean_key_vars = var.groupby( ['year', 'V024'] )[['V201', 'V613', 'V715','V511']].mean().rename(columns = {'V201' : "mean_total_children" , 'V613': "mean_ideal_children" , 
                                                                                                                   'V715': "mean_hb_yr_educ", 'V511': "mean_first_marriage"}).reset_index()

# We call "mean_key_vars" to see our result 
mean_key_vars.head()

year          V024  mean_total_children  mean_ideal_children  \
0  2019       Huanuco             1.989754             2.446940   
1  2019           Ica             1.671034             2.353429   
2  2019         Junin             1.836656             2.199660   
3  2019   La Libertad             1.704082             2.355519   
4  2019    Lambayeque             1.673928             2.399849   

   mean_hb_yr_educ  mean_first_marriage  
0         8.842528            19.818632  
1        11.825332            20.747208  
2        10.942122            20.683761  
3         9.907909            20.774678  
4        11.123396            20.953831

8. Reshape `mean_key_vars` from wide to long. We want a dataframe with three columns **dpto, variables, values**. Name this object as `reshape_mean_key_vars`. **Hint: Use melt method**.

In [106]:
#In this case we will use "melt method" to generate a dataframe with the three columns requested
reshape_mean_key_vars = pd.melt(mean_key_vars, id_vars = ['V024'] , value_vars= ["mean_total_children", "mean_ideal_children", "mean_hb_yr_educ" , "mean_first_marriage"], 
                               ).rename(columns = {'V024': 'dpto'})

reshape_mean_key_vars

dpto             variable      value
0        Huanuco  mean_total_children   1.989754
1            Ica  mean_total_children   1.671034
2          Junin  mean_total_children   1.836656
3    La Libertad  mean_total_children   1.704082
4     Lambayeque  mean_total_children   1.673928
..           ...                  ...        ...
95      Ayacucho  mean_first_marriage  20.340228
96     Cajamarca  mean_first_marriage  19.610695
97        Callao  mean_first_marriage  21.517241
98         Cusco  mean_first_marriage  20.283154
99  Huancavelica  mean_first_marriage  19.780769

[100 rows x 3 columns]

9. Replicate your answers for questions 7 and 8, but in one line of code. Make it the most simple as possible. **NO HINT**

10. Merge `reshape_mean_key_vars` with `endes_2019`. Name this object `final_result`. **Hint: Use merge.**

In [ ]:
#se observa el nombre de las columnas de la primera base de datos
reshape_mean_key_vars.columns

In [ ]:
#se observa el nombre de las columnas de la segunda base de datos
endes_2019.columns
#se realiza un rename para  poder hacer el merge
endes_2019.rename(columns={'V007': 'year', 'V024': 'dpto'}, inplace=True)
#se constata que se haya cambiado el nombre a las variables
endes_2019.columns

In [ ]:
#Se realiza el merge de los Data frames
final_result = pd.merge(endes_2019, reshape_mean_key_vars, on=['dpto'], how='left')
#se observa el resultado
final_result